keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,774 @@
+implementations is required, by introducing an intermediate, node-level abstraction provider. This
+abstraction is expected to be the result of applying Kubernetes' `Service` semantics and business
+logic to a simpler, more stable API.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- provide a node-level abstraction of the cluster-wide `Service` semantics through an API
+- allow easier, more stable proxy implementations that don't need updates when `Service` business
+  logic changes
+- provide a client library with minimal dependencies
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- (optional) help proxy implementations using the same subsystem (ie iptables) to cooperate more
+  easily
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- provide equivalent implementations of third-party ones
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible as usual via
+TCP (`127.0.0.1:12345`) and/or via a socket (`unix:///path/to/proxy.sock`).
+
+- it will connect to the API server and watch resources, like the current proxy;
+- then, it will process them, applying Kubernetes specific business logic like topology computation
+  relative to the local host;
+- finally, provide the result of this computation to client via a gRPC watchable API.
+
+This decoupling allows kube-proxy and implementation to evolve in their own timeframes. For
+instance, introducing optimizations like EndpointSlice or new business semantics like Topology
+does not trigger a rebuild/release of any proxy implementation.
+
+The idea is to send the full state to the client, so implementations don't have to do
+diff-processing and maintain any internal state. This should provide simple implementations,
+reliable results and still be quite optimal, since many kernel network-level objects are
+updated via atomic replace APIs. It's also a protection from slow readers, since no stream has to
+be buffered.
+
+Since the node-local state computed by the new proxy will be simpler and node-specific, it will
```

Comment:
Can you help me to understand what do you mean by node-specific, each service/endpoint change will affect all nodes (unless topology is implemented), or is not about that?

Reply:
yes, without topology there could have been other places to put this API, like the api-server, but since topology is desired, the mistake of *not* putting this as a per-node instance have not been made here :smile: